go/ephemeral/device_ek_storage.go

```diff
@@ -189,3 +212,55 @@ func (s *DeviceEKStorage) MaxGeneration(ctx context.Context) (maxGeneration keyb
 	}
 	return maxGeneration, nil
 }
+
+func (s *DeviceEKStorage) DeleteExpired(ctx context.Context) (expired []keybase1.EkGeneration, err error) {
+	defer s.G().CTrace(ctx, "DeviceEKStorage#DeleteExpired", func() error { return err })()
+	s.Lock()
+	defer s.Unlock()
+
+	cache, err := s.getCache(ctx)
+	if err != nil {
+		return expired, err
+	}
+
+	keyMap := make(keyExpiryMap)
+	for generation, deviceEK := range cache {
+		keyMap[generation] = deviceEK.Metadata.Ctime
+	}
+
+	latestMerkleRoot, err := s.G().GetMerkleClient().FetchRootFromServer(ctx, libkb.EphemeralKeyMerkleFreshness)
+	if err != nil {
+		return expired, err
+	}
+	expired = getExpiredGenerations(keyMap, keybase1.Time(latestMerkleRoot.Ctime()))
+	epick := libkb.FirstErrorPicker{}
+	for _, generation := range expired {
+		epick.Push(s.delete(ctx, generation))
+	}
+
+	epick.Push(s.deletedWrongEldestSeqno(ctx))
+	return expired, epick.Error()
+}
+
+func (s *DeviceEKStorage) deletedWrongEldestSeqno(ctx context.Context) (err error) {
+	defer s.G().CTrace(ctx, "DeviceEKStorage#deletedWrongEldestSeqno", func() error { return err })()
+	keys, err := s.storage.AllKeys(ctx)
+	if err != nil {
+		return err
+	}
+	uv, err := getCurrentUserUV(ctx, s.G())
+	if err != nil {
+		return err
+	}
+	epick := libkb.FirstErrorPicker{}
+	for _, key := range keys {
+		eldestSeqno, err := s.keyToEldestSeqno(key)
+		if err != nil || eldestSeqno < 0 {
+			continue
```

Comment:
This is an unexpected error, right? Probably we should at least log it. With unexpected errors, I do lean towards propagating them. Though with errors that occur during destruction/deletion/cleanup, I also lean towards cleaning up as much as possible despite errors. So I don't feel strongly apart from that we should log it.